pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/PslSourceOffset.java

```diff
@@ -16,72 +16,59 @@
 import com.google.cloud.pubsublite.Offset;
 import com.google.cloud.pubsublite.Partition;
-import com.google.common.collect.ImmutableList;
+import com.google.common.base.Objects;
 import com.google.common.collect.ImmutableMap;
-import java.io.IOException;
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.stream.Collectors;
 
-public final class PslSourceOffset extends org.apache.spark.sql.sources.v2.reader.streaming.Offset {
-  private static final ObjectMapper objectMapper =
-      new ObjectMapper().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
+public class PslSourceOffset {
 
   private final ImmutableMap<Partition, Offset> partitionOffsetMap;
 
   public PslSourceOffset(Map<Partition, Offset> map) {
     this.partitionOffsetMap = ImmutableMap.copyOf(map);
   }
 
-  public static PslSourceOffset merge(PslSourceOffset o1, PslSourceOffset o2) {
-    Map<Partition, Offset> result = new HashMap<>(o1.partitionOffsetMap);
-    o2.partitionOffsetMap.forEach(
-        (k, v) -> result.merge(k, v, (v1, v2) -> Collections.max(ImmutableList.of(v1, v2))));
-    return new PslSourceOffset(result);
+  public Map<Partition, Offset> getPartitionOffsetMap() {
+    return ImmutableMap.copyOf(partitionOffsetMap);
   }
 
-  public static PslSourceOffset merge(PslPartitionOffset[] offsets) {
-    Map<Partition, Offset> map = new HashMap<>();
-    for (PslPartitionOffset po : offsets) {
-      assert !map.containsKey(po.partition()) : "Multiple PslPartitionOffset has same partition.";
-      map.put(po.partition(), po.offset());
-    }
-    return new PslSourceOffset(map);
+  public static Builder newBuilder(long partitionCount) {
+    return new Builder(partitionCount);
   }
 
-  public static PslSourceOffset fromJson(String json) {
-    Map<Long, Long> map;
-    try {
-      map = objectMapper.readValue(json, new TypeReference<Map<Long, Long>>() {});
-    } catch (IOException e) {
-      throw new IllegalStateException("Unable to deserialize PslSourceOffset.", e);
+  public static class Builder {
+
+    private final Map<Partition, Offset> partitionOffsetMap = new HashMap<>();
+
+    private Builder(long partitionCount) {
+      for (long i = 0; i < partitionCount; i++) {
+        partitionOffsetMap.put(Partition.of(i), Offset.of(0));
+      }
+    }
+
+    public Builder set(Partition partition, Offset offset) {
+      partitionOffsetMap.replace(partition, offset);
+      return this;
+    }
+
+    public PslSourceOffset build() {
+      return new PslSourceOffset(partitionOffsetMap);
     }
-    Map<Partition, Offset> partitionOffsetMap =
-        map.entrySet().stream()
-            .collect(Collectors.toMap(e -> Partition.of(e.getKey()), e -> Offset.of(e.getValue())));
-    return new PslSourceOffset(partitionOffsetMap);
   }
 
-  public Map<Partition, Offset> getPartitionOffsetMap() {
-    return this.partitionOffsetMap;
+  @Override
```

Comment:
instead of implementing these, make this an autovalue class and use an autovalue builder. That generates equality and hashcode implementations for you.

Reply:
I intended to use customized builder for convenience that autovalue doesn't support and I autogenerated those equal hashcode from intelliJ, but that convenience it not that much. I instead changed the call site with custom logic now.